### DISCUSSION TITLE
brew install upgrades everything

### DISCUSSION THREAD

**Discussion by chasrmartin** [Homebrew/discussions#57] [Everyday usage]
This is new behavior. If I do `brew install x`, brew first does `brew update` — which makes some sense — then installs the new package. But then it upgrades everything else.

Often I don't *want* to upgrade everything else, either because I have code using the current versions, or because I just don't want to spend the time to upgrade *everything* which involves a lot of network traffic and sometimes lengthy compilations.

Worse yet is what I just observed: I did `brew install nmap`. After installing nmap, brew proceeded to update a whole bunch of other things, including updating erlang. Or rather, trying to, because the erlang update failed. the `erl` command works, at least, but it's not immediately obvious if there is something else broken; I'll need to do regression tests on my erlang code.

This particular change in behavior is a mistake that should be corrected. What's more, this isn't the first time the behavior of the brew command has changed by surprise. It's really hard to use brew in any larger project we one can't depend that it will work the same way.

**Comment by iMichka** [comment]
`brew deps nmap` shows that this package depends on `openssl@1.1`. Openssl was updated to version 1.1.1h, to provide you the latest security fixes.

`brew uses openssl@1.1` shows that `erlang` depends on `openssl@1.1` too. So we update that package too. It's called a reverse dependency.

What would you expect? Leaving a broken `erlang` package is not an option.

And this is not a change in behaviour: as far as I recall, we have done it this way for years now.

**Reply by pkoziol** [reply]
I was also surprised by this, but I guess it makes sense. I would love to have brew asking for confirmation though.

> Hey, _installing that-random-project-you-wanted-to-try-out_ will update _openssl_, which will also update _various-important-things-like-ansible-or-something_.
> Do you want to continue? [y/n]

**Reply by danielnachun** [reply]
If you need to hold a package at a certain version, you can use `brew pin`.  Note that if that package is a dependency for any other packages, those packages cannot be upgraded either.  This behavior is intentional as we only test packages with the latest version of their dependencies.  You can also use `brew tap` and `brew extract` like is shown here: https://github.com/Homebrew/discussions/discussions/155#discussioncomment-133894 to extract a copy of a formula at a version you want into your own tap.  Doing it this way allows you keep that version around forever, though you may eventually need to extract some of the dependencies of that formula if they become too new to support it.

**Reply by PhillyOne** [reply]
This feature is a pain. I stopped using my laptop for 2 months and installed a single package that pretty much forced my whole system to get updated with new versions of everything. Other package managers warn you before proceeding. This "feature" is catching many developers off guard. For instance, I installed the package "jq". I have many applications dependant on the current version of "php" I have installed. Brew has gone off and now decided that I need php8, python 3.8, broken my curl implementation and a number of other things. All because I wanted a simple package. Homebrew is like the only reason I still own a mac. If this becomes a pain it's the nail in the coffin for me. I shouldn't have to pin everything. I don't have this issue on my redhat/centos systems when upgrading via yum. Python 3.8 might be newer and so is php8 but just because its logically correct to do that, it doesn't mean it should be done. I should be warned before major dependencies like these are updated. In practice, this is just an inconvenience. There needs to be some kind of option to stop this from happening automatically without having to pin everything. This just created me hours of unnecessary work.

**Reply by SMillerDev** [reply]
Homebrew is an evergreen package manager by design and works best if you update often, which is the opposite idea from CentOS/RedHat. You'd have the same issue if you used an Arch based system, which shares Homebrew's philosophy.

> I have many applications dependant on the current version of "php" I have installed.

You can use `brew-php-switcher` to switch versions. All supported PHP versions are available in brew.

> I should be warned before major dependencies like these are updated.

There are feeds for that, for example https://www.php.net/archive/2020.php#2020-11-26-3 warned you that the latest version was now PHP 8.

**Reply by PhillyOne** [reply]
As I said, I went to install a single package and it broke a ton of other things on my system (some of which I don't maintain). The direction this is going will have me removing this off my system altogether and just use a VM. As a developer, the ground should not constantly be moving beneath my feet. That might be the philosophy, but it doesn't benefit any developer who needs consistency. Homebrew shouldn't become a bottleneck in my development process. I came to develop a 15 minute feature for my client now i'm wasting hours trying to fix dependency issues with brew. I just don't know in what world that makes sense.

**Reply by tlk** [reply]
> There needs to be some kind of option to stop this from happening automatically without having to pin everything.

I agree with you and have found that setting the environment variable `HOMEBREW_NO_AUTO_UPDATE` makes homebrew more pleasant to use than with its default settings. Search this discussion forum for caveats as Sean and I have discussed this before.

This will skip the automatic `brew update` step so you are in control of when you want to fetch formula updates. Because of that, you can put your laptop away for 2 months, come back and run `brew install jq` and with pretty high confidence get something that works, although jq may not be the most recently released version. If it doesn't work you will have to run `brew update` and proceed as before - only this time you will get the most recent release of jq and a bunch of other packages will be upgraded (and homebrew itself).

**Reply by SMillerDev** [reply]
Just be aware that setting it can cause things like: https://github.com/Homebrew/discussions/discussions/2035

**Reply by psifertex** [reply]
I was stuck tethered to my rate-limited cellphone connection phone on laptop battery trying to install a single package when suddenly every package started to upgrade. This is such an awful anti-pattern I can't even begin to express it. Trying to kill it mid-install is a terrible choice, letting it continue was a terrible choice.

This should absolutely not be default behavior in any sane package manager. Warn, ask, refuse to install the original package if there are conflicts, any of these are fine options. 

The current behavior is not acceptable.

**Reply by alebcay** [reply]
> The current behavior is not acceptable.

Your ardor is noted. I would encourage you to be the change you want to see in the world and open a pull request with the proposed changes at https://github.com/Homebrew/brew.

Because, frankly, the status quo works for Homebrew maintainers (at least enough that nobody has really tried to implement a change) and it seems to work for [a lot of other users too](https://formulae.brew.sh/analytics/). I imagine that this change would not be trivial and so someone who is not adversely affected by the current behavior might be hard-pressed to really change it (and, more importantly, test it).

It's evident that this behavior is not what you or some other users are expecting, and it's fairly clear, based on this thread and other discussions, what the desired behavior is. However, we can discuss the desired behavior all day but your experience with Homebrew will remain just as miserable as nobody is really inclined to implement it.

And at worst, if such a pull request is not accepted, now you have a fork of Homebrew that does do what you want it to do.

On the other hand, if nobody wants to make/accept the proposed change (and you don't want to have your own version of Homebrew), maybe its time to consider a different package manager or tool for your use cases. No tool or package manager can support every usecase (except `emacs`, perhaps). After all, Homebrew is only one of several package managers on the street supporting macOS; they have all been designed with different use cases in mind and are staffed by some fine folks too (I've had the chance to interact with them on occasion).

Now, more specifically with regard to your proposed alternate behaviors:

> Warn, ask, refuse to install the original package if there are conflicts, any of these are fine options.

I fail to see how "warn" is really helpful here - Homebrew would just print a warning and proceed to upgrade everything? The only way out for the user would be to interrupt anyways, and like you mentioned, that is hardly ideal.

The latter two would break existing user flows and as such, in my opinion, would not be fine options. Even with deprecation notices and some warning, such changes inevitably bring users' whose automation broke on a `brew update` to come crying to us and screaming that we change it back.

Frankly, there are no fine options (the status quo included) - only bad and worse ones.

**Reply by carlocab** [reply]
> This should absolutely not be default behavior in any sane package manager. Warn, ask, refuse to install the original package if there are conflicts, any of these are fine options.

The behaviour you propose is not acceptable. This makes `brew` unusable for scripting and automation in, for example, CI environments, which form a significant part of Homebrew's user base.

Understand that Homebrew's default behaviour cannot just be optimised for your personal use-case, because there are many other users who use it differently, and changing this is likely to break these users. This would absolutely be a terrible thing to do for any sane piece of software. And, frankly, I find the suggestion to do so and the tone in which that suggestion was made disgustingly self-centred.

If you don't like the default behaviour, you can set either (or both) of `HOMEBREW_NO_AUTO_UPDATE` or `HOMEBREW_NO_INSTALL_UPGRADE`.

**Reply by psifertex** [reply]
I've since done both -- however as this was not previously the default, it seems like poor behavior. 

If I run `brew upgrade` I expect to get all packages. If I run `brew install $PKG` I am absolutely not expecting to get all packages upgraded. Most of the time it's probably just a minor annoyance (which is why you don't have a flood of people finding this discussion). However, in some instances it's a huge headache such as my situation yesterday. 

Ironically, soon after making my comment above I was working on some of our CI infrastructure and was bitten by the _exact same bug_ again (╯°□°）╯︵ ┻━┻  

Which is especially funny given your comment that it's intended for CI!

I needed a new version of `mkdocs` and we'd (unfortunately) installed it using `brew` instead of `pip`. When I upgraded mkdocs what I instead got was a new python. Which required manually installing a bunch of other python dependencies that other parts of the CI depended on as there was now a totally different python3 version installed! 

I fail to see the argument how this is _better_ "because CI".

**Reply by psifertex** [reply]
@alebcay Are you actually proposing that if I make a PR to change this default behavior it would be accepted? That sounds very unlikely given the other comments to the effect that HB devs prefer this behavior! 

I'm merely trying to point out that I think the default behavior should be more "do what the user asks for". I think you're violating the principle of least surprise pretty badly here.

**Reply by carlocab** [reply]
> I am absolutely not expecting to get all packages upgraded.

Neither am I. That's not what's happening. What happens when you `brew install foo` is that everything connected to `foo` via the dependency graph is updated.

> Which required manually installing a bunch of other python dependencies that other parts of the CI depended on as there was now a totally different python3 version installed!

This should only have upgraded you to a new patch version of Python3 (unless it's been ages since you did `brew update`/`brew upgrade`). In any case, this is probably informative: https://justinmayer.com/posts/homebrew-python-is-not-for-you/

> I fail to see the argument how this is _better_ "because CI".

CI was just one example, so reducing what I said to "because CI" is a pretty awful straw man.

> Are you actually proposing that if I make a PR to change this default behavior it would be accepted? That sounds very unlikely given the other comments to the effect that HB devs prefer this behavior!

It might be unlikely, but we won't know unless you open a PR. Perhaps there's an even better solution we can hash out of a PR.

> I think you're violating the principle of least surprise pretty badly here.

Maybe, but, again. What is least surprising is different for different users.

**Reply by danielnachun** [reply]
The important thing to remember here is that when we have to update the dependency graph, it is usually to avoid having some packages which depend on one version of Python, for example, and other packages which depend on a different version of Python.  This "dependency hell" is a huge headache for both users and maintainers, and package ecosystems like PyPI, NPM and Cabal which use explicit versioning for their dependencies struggle with this issue all the time.  

>Which required manually installing a bunch of other python dependencies that other parts of the CI depended on as there was now a totally different python3 version installed!

Homebrew is not an environment/library manager - it is more similar to a package manager for a rolling release Linux distribution like Arch Linux or Gentoo.  This sounds like a use case for `pyenv` or Anaconda, which are designed to function as environment or library managers.  

There have been proposals for ways to implement a Homebrew equivalent of environments, but there are some technical obstacles that would have to be overcome for that to be possible.  And if such a system is implemented, that would not change the policy in homebrew-core of only supporting the latest version of formulae.  We do have a small number of formulae for which we support multiple major versions because the upstream developer backports fixes (see `mariadb` for example), and I have found that the more versions we have to support, the more work it creates for us.  Trying to support versions of formulae which are not supported by their upstream developer would be far more difficult.

**Reply by psifertex** [reply]
Again, all I ask is that you give the user more information or options before you do this. Some otherwise simple packages result in massive dependency chains that the user absolutely would not choose to upgrade given the option in many situations. 

I'm not arguing about fundamental design philosophy here. Just to make the defaults less... 💣. Given the choice I would have said no while on limited bandwidth/battery and would have installed that tool I needed via some other mechanism or just waited. Let automation use an environment variable override like HOMEBREW_ALWAYS_UPGRADE and let the default prompt the user "hey, you've requested to install this package, but to do that I'm going to need to upgrade these 30 other packages...".

**Reply by SMillerDev** [reply]
> I'm not arguing about fundamental design philosophy here.

... proceeds to argue about fundamental design philosophies.

I realize it sucks for you on a metered connection, but is brew really responsible for your lack of knowledge about it? If you read the FAQ at some point you would have known this. If you checked earlier updates you would have known this.

**Reply by psifertex** [reply]
Are you seriously claiming that asking for a warning message is asking for a change to a core design philosophy?

**Reply by psifertex** [reply]
FYI -- I just double checked the FAQ and saw no references to `brew install` upgrading other packages.

It talks about [pinning](https://docs.brew.sh/FAQ#how-do-i-stop-certain-formulae-from-being-updated) and it talks about how `brew upgrade` can [upgrade other things](https://docs.brew.sh/FAQ#why-does-brew-upgrade-formula-also-upgrade-a-bunch-of-other-stuff) (which is far more intuitive than "I'm installing a new package and it will upgrade other packages" which is the main thread we're discussing here).

Did I miss a different answer?

**Reply by cen1** [reply]
Just installed nmap, ended up with upgraded Postgres database. This is the most ridiculous "feature" ever. No prompt, just full speed ahead essentially nuking the entire system.

**Reply by ghost** [reply]
> Just installed nmap, ended up with upgraded Postgres database. This is the most ridiculous "feature" ever. No prompt, just full speed ahead essentially nuking the entire system.

Yeah exactly

**Reply by ihuk** [reply]
Who thought this was a good idea? I installed some stupid utility and it went ahead and basically did a system update. And why would it upgrade Postgres!?

**Reply by guptamitesh** [reply]
Developer has taken on his ego it seems. Seriously just give a warning its going to change so much and I will live happily with whatever version I had previously.
